// Spiders.js/src/ChannelManager.js
const commMedium_1 = require("./commMedium");
/**
 * Created by flo on 18/01/2017.
 */
var utils = require("./utils");
class ChannelManager extends commMedium_1.CommMedium {
    init(messageHandler) {
        super.init(messageHandler);
        this.connections = new Map();
    }
    newConnection(actorId, channelPort) {
        this.connections.set(actorId, channelPort);
        channelPort.onmessage = (ev) => {
            this.messageHandler.dispatch(JSON.parse(ev.data), ev.ports);
        };
    }
    //Open connection to Node.js instance owning the object to which the far reference refers to
    openConnection(actorId, actorAddress, actorPort) {
        this.socketHandler.openConnection(actorId, actorAddress, actorPort);
    }
    hasConnection(actorId) {
        var inChannel = this.connections.has(actorId);
        var connected = this.connectedActors.has(actorId);
        var disconnected = this.socketHandler.disconnectedActors.indexOf(actorId) != -1;
        return inChannel || connected || disconnected;
    }
    sendMessage(actorId, message, first = true) {
        if (this.connections.has(actorId)) {
            this.connections.get(actorId).postMessage(JSON.stringify(message));
        }
        else if (this.connectedActors.has(actorId) || this.socketHandler.disconnectedActors.indexOf(actorId) != -1) {
            this.socketHandler.sendMessage(actorId, message);
        }
        else {
            //Dirty, but it could be that an actor sends a message to the application actor, leading it to spawn a new actor and returning this new reference.
            //Upon receiving this reference the spawning actor immediatly invokes a method on the reference, but hasn't received the open ports message
            if (first) {
                var that = this;
                setTimeout(() => {
                    that.sendMessage(actorId, message, false);
                }, 10);
            }
            else {
                throw new Error("Unable to send message to unknown actor (channel manager): " + actorId + " in : " + this.messageHandler.thisRef.ownerId);
            }
        }
    }
}
exports.ChannelManager = ChannelManager;
